examples/social/modules/shared_data/src/main/scala/com/my/metagraph_social/shared_data/calculated_state/CalculatedStateService.scala

```diff
@@ -0,0 +1,61 @@
+package com.my.metagraph_social.shared_data.calculated_state
+
+import cats.effect.{Async, Ref}
+import cats.syntax.all._
+import com.my.metagraph_social.shared_data.types.States.SocialCalculatedState
+import io.circe.Json
+import io.circe.syntax.EncoderOps
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.security.hash.Hash
+
+import java.nio.charset.StandardCharsets
+
+trait CalculatedStateService[F[_]] {
+  def get: F[CalculatedState]
+
+  def set(
+    snapshotOrdinal: SnapshotOrdinal,
+    state          : SocialCalculatedState
+  ): F[Boolean]
+
+  def hash(
+    state: SocialCalculatedState
+  ): F[Hash]
+}
+
+object CalculatedStateService {
+  def make[F[_] : Async](
+    externalStorageService: ExternalStorageService[F]
+  ): F[CalculatedStateService[F]] =
+    Ref.of[F, CalculatedState](CalculatedState.empty).map { stateRef =>
+      new CalculatedStateService[F] {
+        override def get: F[CalculatedState] = stateRef.get
+
+        override def set(
+          snapshotOrdinal: SnapshotOrdinal,
+          state          : SocialCalculatedState
+        ): F[Boolean] = for {
+          _ <- externalStorageService.set(snapshotOrdinal, state)
```

Comment:
I do not believe this is an atomic operation at the moment, so the external store may fail to update while the internal memory is updated (thus getting out of sync. I would need to double-check but I think you want to use `stateRef.modify` and to have the storage update side-effect as part of that modification. This would ensure that the operation is atomic, like what is done [here](https://github.com/scasplte2/dilf4s/blob/5bc73bd53f703e76a96a39a081574fc6e14c3525/modules/root/src/main/scala/xyz/kd5ujc/storage/interpreters/versioned_store/package.scala#L57)

Reply:
I think the atomic version would be
```
 override def set(
          snapshotOrdinal: SnapshotOrdinal,
          state          : SocialCalculatedState
        ): F[Boolean] = stateRef.modify { currentState =>
        val effect = externalStorageService.set(snapshotOrdinal, state)
        val currentCalculatedState = currentState.state
            val updatedUsers = state.users.foldLeft(currentCalculatedState.users) {
              case (acc, (address, value)) =>
                acc.updated(address, value)
            }
         val state = CalculatedState(snapshotOrdinal, SocialCalculatedState(updatedUsers))
         
         (state, effect)
      }.flatten
        .as(true)
        .handleErrorWith { err =>
          Logger[F].warn(s"Versioned store update failed with error: $err") *> false.pure[F]
        }
```

Reply:
Updated